File: src/cppinterface/cb_bundlediagonaltrustregionprox.jl
@doc raw"""
    CBBundleDiagonalTrustRegionProx(Din::CBMatrix, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1)

initialize to this diagonal matrix
"""
CBBundleDiagonalTrustRegionProx(Din::CBMatrix, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1) = CBBundleDiagonalTrustRegionProx(@ccall libcb.cb_bundlediagonaltrustregionprox_new(Din.data::Ptr{Cvoid}, (isnothing(vp) ? C_NULL : vp.data)::Ptr{Cvoid}, local_scaling::Cint, bounds_scaling::Cint, cbinc::Cint)::Ptr{Cvoid})

@doc raw"""
    CBBundleDiagonalTrustRegionProx(dim::Integer, d::Real, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1)

initialize to a diaognal matrix d*identity of dimesion dim
"""
CBBundleDiagonalTrustRegionProx(dim::Integer, d::Real, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1) = CBBundleDiagonalTrustRegionProx(@ccall libcb.cb_bundlediagonaltrustregionprox_new2(dim::Cint, d::Cdouble, (isnothing(vp) ? C_NULL : vp.data)::Ptr{Cvoid}, local_scaling::Cint, bounds_scaling::Cint, cbinc::Cint)::Ptr{Cvoid})

@doc raw"""
    CBBundleDiagonalTrustRegionProx(dim::Integer = 0, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1)

initialize to a zero diaognal matrix of dimesion dim
"""
CBBundleDiagonalTrustRegionProx(dim::Integer = 0, vp::Union{<:CBVariableMetricSelection,Nothing} = nothing, local_scaling::Bool = false, bounds_scaling::Bool = false, cbinc::Integer = -1) = CBBundleDiagonalTrustRegionProx(@ccall libcb.cb_bundlediagonaltrustregionprox_new3(dim::Cint, (isnothing(vp) ? C_NULL : vp.data)::Ptr{Cvoid}, local_scaling::Cint, bounds_scaling::Cint, cbinc::Cint)::Ptr{Cvoid})

@doc raw"""
    cb_set_weightu!(self::CBBundleDiagonalTrustRegionProx, in_weightu::Real)

set the weight of the proximal term
"""
cb_set_weightu!(self::CBBundleDiagonalTrustRegionProx, in_weightu::Real) = @ccall libcb.cb_bundlediagonaltrustregionprox_set_weightu(self.data::Ptr{Cvoid}, in_weightu::Cdouble)::Cvoid

@doc raw"""
    cb_get_weightu(self::CBBundleDiagonalTrustRegionProx)

returns the current weight of the proximal term
"""
cb_get_weightu(self::CBBundleDiagonalTrustRegionProx) = @ccall libcb.cb_bundlediagonaltrustregionprox_get_weightu(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_term_corr(self::CBBundleDiagonalTrustRegionProx)

returns a correction factor for termination precision if the quadratic term is strong
"""
cb_get_term_corr(self::CBBundleDiagonalTrustRegionProx) = @ccall libcb.cb_bundlediagonaltrustregionprox_get_term_corr(self.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_D(self::CBBundleDiagonalTrustRegionProx)

returns the diagonal D of the diagonal scaling matrix
"""
cb_get_D(self::CBBundleDiagonalTrustRegionProx) = (@ccall libcb.cb_bundlediagonaltrustregionprox_get_d(self.data::Ptr{Cvoid})::Ptr{Cvoid}; return self)

@doc raw"""
    cb_set_D!(self::CBBundleDiagonalTrustRegionProx, in_D::CBMatrix)

set the diagonal (it needs to be >=0 but this is not checked)
"""
cb_set_D!(self::CBBundleDiagonalTrustRegionProx, in_D::CBMatrix) = @ccall libcb.cb_bundlediagonaltrustregionprox_set_d(self.data::Ptr{Cvoid}, in_D.data::Ptr{Cvoid})::Cvoid

@doc raw"""
    cb_dim(self::CBBundleDiagonalTrustRegionProx)

returns the dimension of the diagonal
"""
cb_dim(self::CBBundleDiagonalTrustRegionProx) = @ccall libcb.cb_bundlediagonaltrustregionprox_dim(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    Base.getindex(self::CBBundleDiagonalTrustRegionProx, i::Integer)

return the i-th element of the diagonal matrix D
"""
Base.getindex(self::CBBundleDiagonalTrustRegionProx, i::Integer) = @ccall libcb.cb_bundlediagonaltrustregionprox_get(self.data::Ptr{Cvoid}, i::Cint)::Cdouble

@doc raw"""
    cb_norm_sqr(self::CBBundleDiagonalTrustRegionProx, B::CBMatrix)

returns \f$\|B\|^2_H\f$
"""
cb_norm_sqr(self::CBBundleDiagonalTrustRegionProx, B::CBMatrix) = @ccall libcb.cb_bundlediagonaltrustregionprox_norm_sqr(self.data::Ptr{Cvoid}, B.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_dnorm_sqr(self::CBBundleDiagonalTrustRegionProx, B::CBMinorantPointer)

returns \f$\|B\|^2_{H^{-1}}\f$
"""
cb_dnorm_sqr(self::CBBundleDiagonalTrustRegionProx, B::CBMinorantPointer) = @ccall libcb.cb_bundlediagonaltrustregionprox_dnorm_sqr(self.data::Ptr{Cvoid}, B.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_is_DLR(self::CBBundleDiagonalTrustRegionProx)

return true if H is of the form diagonal matrix plus Gram matrix of a low rank matrix
"""
cb_is_DLR(self::CBBundleDiagonalTrustRegionProx) = Bool(@ccall libcb.cb_bundlediagonaltrustregionprox_is_dlr(self.data::Ptr{Cvoid})::Cint)

@doc raw"""
    cb_add_H(self::CBBundleDiagonalTrustRegionProx, big_sym::CBSymmatrix, start_index::Integer = 0)

add H to the dense symmetric matrix as a principal submatrix starting at position start_index
"""
cb_add_H(self::CBBundleDiagonalTrustRegionProx, big_sym::CBSymmatrix, start_index::Integer = 0) = @ccall libcb.cb_bundlediagonaltrustregionprox_add_h(self.data::Ptr{Cvoid}, big_sym.data::Ptr{Cvoid}, start_index::Cint)::Cint

@doc raw"""
    cb_add_Hx(self::CBBundleDiagonalTrustRegionProx, x::CBMatrix, outplusHx::CBMatrix, alpha::Real = 1.)

adds \f$alpha*Hx\f$ to outplusHx and returns this
"""
cb_add_Hx(self::CBBundleDiagonalTrustRegionProx, x::CBMatrix, outplusHx::CBMatrix, alpha::Real = 1.) = (@ccall libcb.cb_bundlediagonaltrustregionprox_add_hx(self.data::Ptr{Cvoid}, x.data::Ptr{Cvoid}, outplusHx.data::Ptr{Cvoid}, alpha::Cdouble)::Ptr{Cvoid}; return self)

@doc raw"""
    cb_apply_Hinv(self::CBBundleDiagonalTrustRegionProx, x::CBMatrix)

returns \f$H^{-1}x\f$
"""
cb_apply_Hinv(self::CBBundleDiagonalTrustRegionProx, x::CBMatrix) = (@ccall libcb.cb_bundlediagonaltrustregionprox_apply_hinv(self.data::Ptr{Cvoid}, x.data::Ptr{Cvoid})::Ptr{Cvoid}; return self)

@doc raw"""
    cb_compute_QP_costs!(self::CBBundleDiagonalTrustRegionProx, Q::CBSymmatrix, d::CBMatrix, constant_minorant::CBMinorantPointer, bundle::CBMinorantBundle, y::CBMatrix, groundset_minorant::CBMinorantPointer, yfixed::Union{<:CBIndexmatrix,Nothing})

computes the dual QP costs Q, d, and the constant offset to the bundle subproblem, see BundleProxObject::compute_QP_costs
"""
function cb_compute_QP_costs!(self::CBBundleDiagonalTrustRegionProx, Q::CBSymmatrix, d::CBMatrix, constant_minorant::CBMinorantPointer, bundle::CBMinorantBundle, y::CBMatrix, groundset_minorant::CBMinorantPointer, yfixed::Union{<:CBIndexmatrix,Nothing})
    offset = Ref{Float64}()
    @ccall libcb.cb_bundlediagonaltrustregionprox_compute_qp_costs(self.data::Ptr{Cvoid}, Q.data::Ptr{Cvoid}, d.data::Ptr{Cvoid}, offset::Ref{Float64}, constant_minorant.data::Ptr{Cvoid}, bundle.data::Ptr{Cvoid}, y.data::Ptr{Cvoid}, groundset_minorant.data::Ptr{Cvoid}, (isnothing(yfixed) ? C_NULL : yfixed.data)::Ptr{Cvoid})::Cint
    return offset[]
end

@doc raw"""
    cb_update_QP_costs!(self::CBBundleDiagonalTrustRegionProx, delta_Q::CBSymmatrix, delta_d::CBMatrix, constant_minorant::CBMinorantPointer, bundle::CBMinorantBundle, center_y::CBMatrix, groundset_minorant::CBMinorantPointer, delta_groundset_minorant::CBMinorantPointer, delta_index::CBIndexmatrix, yfixed::Union{<:CBIndexmatrix,Nothing})

updates the dual QP costs Q, d, and the constant offset to the bundle subproblem, see BundleProxObject::update_QP_costs
"""
function cb_update_QP_costs!(self::CBBundleDiagonalTrustRegionProx, delta_Q::CBSymmatrix, delta_d::CBMatrix, constant_minorant::CBMinorantPointer, bundle::CBMinorantBundle, center_y::CBMatrix, groundset_minorant::CBMinorantPointer, delta_groundset_minorant::CBMinorantPointer, delta_index::CBIndexmatrix, yfixed::Union{<:CBIndexmatrix,Nothing})
    delta_offset = Ref{Float64}()
    @ccall libcb.cb_bundlediagonaltrustregionprox_update_qp_costs(self.data::Ptr{Cvoid}, delta_Q.data::Ptr{Cvoid}, delta_d.data::Ptr{Cvoid}, delta_offset::Ref{Float64}, constant_minorant.data::Ptr{Cvoid}, bundle.data::Ptr{Cvoid}, center_y.data::Ptr{Cvoid}, groundset_minorant.data::Ptr{Cvoid}, delta_groundset_minorant.data::Ptr{Cvoid}, delta_index.data::Ptr{Cvoid}, (isnothing(yfixed) ? C_NULL : yfixed.data)::Ptr{Cvoid})::Cint
    return delta_offset[]
end

@doc raw"""
    cb_apply_modification!(self::CBBundleDiagonalTrustRegionProx, gsmdf::CBGroundsetModification)

when BundleSolver is called to modify the groundset it also calls this
"""
cb_apply_modification!(self::CBBundleDiagonalTrustRegionProx, gsmdf::CBGroundsetModification) = @ccall libcb.cb_bundlediagonaltrustregionprox_apply_modification(self.data::Ptr{Cvoid}, gsmdf.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_projected_clone!(self::CBBundleDiagonalTrustRegionProx, indices::CBIndexmatrix)

* @brief in order to allow for fixed variables, this generates a clone restricted to the given indices
   
"""
cb_projected_clone!(self::CBBundleDiagonalTrustRegionProx, indices::CBIndexmatrix) = CBBundleProxObject(@ccall libcb.cb_bundlediagonaltrustregionprox_projected_clone(self.data::Ptr{Cvoid}, indices.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_supports_diagonal_bounds_scaling(self::CBBundleDiagonalTrustRegionProx)

* @brief this implementation supports a diagonal scaling heuristic for
        bounds in the groundset, therefore the following routine has to return true.
     
"""
cb_supports_diagonal_bounds_scaling(self::CBBundleDiagonalTrustRegionProx) = Bool(@ccall libcb.cb_bundlediagonaltrustregionprox_supports_diagonal_bounds_scaling(self.data::Ptr{Cvoid})::Cint)

@doc raw"""
    cb_diagonal_bounds_scaling_update!(self::CBBundleDiagonalTrustRegionProx, D_update::CBMatrix)

* @brief if supported, D_update has to contain nonnegative numbers that
        are permanently added to the diagonal here. It is important to keep
        track of this change only if afterwards update_QP_costs is called before
        compute_QP_costs. In this case the only nonzero enries in D_update must
        be those of delta_index
     
"""
cb_diagonal_bounds_scaling_update!(self::CBBundleDiagonalTrustRegionProx, D_update::CBMatrix) = @ccall libcb.cb_bundlediagonaltrustregionprox_diagonal_bounds_scaling_update(self.data::Ptr{Cvoid}, D_update.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_supports_lowrank_variable_metric(self::CBBundleDiagonalTrustRegionProx)

returns true if dynamic scaling with low rank matrices is supported
"""
cb_supports_lowrank_variable_metric(self::CBBundleDiagonalTrustRegionProx) = Bool(@ccall libcb.cb_bundlediagonaltrustregionprox_supports_lowrank_variable_metric(self.data::Ptr{Cvoid})::Cint)

@doc raw"""
    cb_supports_diagonal_variable_metric(self::CBBundleDiagonalTrustRegionProx)

returns true if dynamic scaling with diagonal matrices is supported
"""
cb_supports_diagonal_variable_metric(self::CBBundleDiagonalTrustRegionProx) = Bool(@ccall libcb.cb_bundlediagonaltrustregionprox_supports_diagonal_variable_metric(self.data::Ptr{Cvoid})::Cint)

@doc raw"""
    cb_apply_variable_metric!(self::CBBundleDiagonalTrustRegionProx, groundset::Union{<:CBVariableMetricModel,Nothing}, model::Union{<:CBVariableMetricModel,Nothing}, aggr::CBMatrix, y_id::Integer, y::CBMatrix, descent_step::Bool, model_maxviol::Real, new_indices::Union{<:CBIndexmatrix,Nothing} = nothing)

see VariableMetric
"""
function cb_apply_variable_metric!(self::CBBundleDiagonalTrustRegionProx, groundset::Union{<:CBVariableMetricModel,Nothing}, model::Union{<:CBVariableMetricModel,Nothing}, aggr::CBMatrix, y_id::Integer, y::CBMatrix, descent_step::Bool, model_maxviol::Real, new_indices::Union{<:CBIndexmatrix,Nothing} = nothing)
    current_weight = Ref{Float64}()
    @ccall libcb.cb_bundlediagonaltrustregionprox_apply_variable_metric(self.data::Ptr{Cvoid}, (isnothing(groundset) ? C_NULL : groundset.data)::Ptr{Cvoid}, (isnothing(model) ? C_NULL : model.data)::Ptr{Cvoid}, aggr.data::Ptr{Cvoid}, y_id::Cint, y.data::Ptr{Cvoid}, descent_step::Cint, current_weight::Ref{Float64}, model_maxviol::Cdouble, (isnothing(new_indices) ? C_NULL : new_indices.data)::Ptr{Cvoid})::Cint
    return current_weight[]
end

@doc raw"""
    cb_add_variable_metric!(self::CBBundleDiagonalTrustRegionProx, diagH::CBMatrix, vecH::CBMatrix)

see BundleProxObject::add_dynamic_scaling()
"""
cb_add_variable_metric!(self::CBBundleDiagonalTrustRegionProx, diagH::CBMatrix, vecH::CBMatrix) = @ccall libcb.cb_bundlediagonaltrustregionprox_add_variable_metric(self.data::Ptr{Cvoid}, diagH.data::Ptr{Cvoid}, vecH.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_push_aft!(self::CBBundleDiagonalTrustRegionProx, aft::Union{<:CBAffineFunctionTransformation,Nothing})

see  BundleProxObject::push_aft();
"""
cb_push_aft!(self::CBBundleDiagonalTrustRegionProx, aft::Union{<:CBAffineFunctionTransformation,Nothing}) = @ccall libcb.cb_bundlediagonaltrustregionprox_push_aft(self.data::Ptr{Cvoid}, (isnothing(aft) ? C_NULL : aft.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_pop_aft!(self::CBBundleDiagonalTrustRegionProx)

see  BundleProxObject::pop_aft();
"""
cb_pop_aft!(self::CBBundleDiagonalTrustRegionProx) = @ccall libcb.cb_bundlediagonaltrustregionprox_pop_aft(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_mfile_data(self::CBBundleDiagonalTrustRegionProx)

output the description of the scaling in mfile-suitable format
"""
cb_mfile_data(self::CBBundleDiagonalTrustRegionProx) = @ccall libcb.cb_bundlediagonaltrustregionprox_mfile_data(self.data::Ptr{Cvoid})::Cint

File: ConicBundle/cppinterface/cb_bundlediagonaltrustregionprox.cpp
dll void cb_bundlediagonaltrustregionprox_destroy(BundleDiagonalTrustRegionProx* self) {
  delete self;
}

dll BundleDiagonalTrustRegionProx* cb_bundlediagonaltrustregionprox_new(const Matrix* Din, VariableMetricSelection* vp = 0, int local_scaling = 0, int bounds_scaling = 0, int cbinc = -1) {
  return new BundleDiagonalTrustRegionProx(*Din, vp, (bool)local_scaling, (bool)bounds_scaling, 0, cbinc);
}

dll BundleDiagonalTrustRegionProx* cb_bundlediagonaltrustregionprox_new2(Integer dim, Real d, VariableMetricSelection* vp = 0, int local_scaling = 0, int bounds_scaling = 0, int cbinc = -1) {
  return new BundleDiagonalTrustRegionProx(dim, d, vp, (bool)local_scaling, (bool)bounds_scaling, 0, cbinc);
}

dll BundleDiagonalTrustRegionProx* cb_bundlediagonaltrustregionprox_new3(Integer dim = 0, VariableMetricSelection* vp = 0, int local_scaling = 0, int bounds_scaling = 0, int cbinc = -1) {
  return new BundleDiagonalTrustRegionProx(dim, vp, (bool)local_scaling, (bool)bounds_scaling, 0, cbinc);
}

dll void cb_bundlediagonaltrustregionprox_set_weightu(BundleDiagonalTrustRegionProx* self, Real in_weightu) {
  self->set_weightu(in_weightu);
}

dll Real cb_bundlediagonaltrustregionprox_get_weightu(const BundleDiagonalTrustRegionProx* self) {
  return self->get_weightu();
}

dll Real cb_bundlediagonaltrustregionprox_get_term_corr(const BundleDiagonalTrustRegionProx* self) {
  return self->get_term_corr();
}

dll const Matrix* cb_bundlediagonaltrustregionprox_get_d(const BundleDiagonalTrustRegionProx* self) {
  return &self->get_D();
}

dll void cb_bundlediagonaltrustregionprox_set_d(BundleDiagonalTrustRegionProx* self, Matrix* in_D) {
  self->set_D(*in_D);
}

dll Integer cb_bundlediagonaltrustregionprox_dim(const BundleDiagonalTrustRegionProx* self) {
  return self->dim();
}

dll Real cb_bundlediagonaltrustregionprox_get(const BundleDiagonalTrustRegionProx* self, Integer i) {
  return (*self)(i);
}

dll Real cb_bundlediagonaltrustregionprox_norm_sqr(const BundleDiagonalTrustRegionProx* self, const Matrix* B) {
  return self->norm_sqr(*B);
}

dll Real cb_bundlediagonaltrustregionprox_dnorm_sqr(const BundleDiagonalTrustRegionProx* self, const MinorantPointer* B) {
  return self->dnorm_sqr(*B);
}

dll int cb_bundlediagonaltrustregionprox_is_dlr(const BundleDiagonalTrustRegionProx* self) {
  return self->is_DLR();
}

dll int cb_bundlediagonaltrustregionprox_add_h(const BundleDiagonalTrustRegionProx* self, Symmatrix* big_sym, Integer start_index = 0) {
  return self->add_H(*big_sym, start_index);
}

dll Matrix* cb_bundlediagonaltrustregionprox_add_hx(const BundleDiagonalTrustRegionProx* self, const Matrix* x, Matrix* outplusHx, Real alpha = 1.) {
  return &self->add_Hx(*x, *outplusHx, alpha);
}

dll Matrix* cb_bundlediagonaltrustregionprox_apply_hinv(const BundleDiagonalTrustRegionProx* self, Matrix* x) {
  return &self->apply_Hinv(*x);
}

dll int cb_bundlediagonaltrustregionprox_compute_qp_costs(BundleDiagonalTrustRegionProx* self, Symmatrix* Q, Matrix* d, Real* offset, const MinorantPointer* constant_minorant, const MinorantBundle* bundle, const Matrix* y, const MinorantPointer* groundset_minorant, Indexmatrix* yfixed) {
  return self->compute_QP_costs(*Q, *d, *offset, *constant_minorant, *bundle, *y, *groundset_minorant, yfixed);
}

dll int cb_bundlediagonaltrustregionprox_update_qp_costs(BundleDiagonalTrustRegionProx* self, Symmatrix* delta_Q, Matrix* delta_d, Real* delta_offset, const MinorantPointer* constant_minorant, const MinorantBundle* bundle, const Matrix* center_y, const MinorantPointer* groundset_minorant, const MinorantPointer* delta_groundset_minorant, const Indexmatrix* delta_index, Indexmatrix* yfixed) {
  return self->update_QP_costs(*delta_Q, *delta_d, *delta_offset, *constant_minorant, *bundle, *center_y, *groundset_minorant, *delta_groundset_minorant, *delta_index, yfixed);
}

dll int cb_bundlediagonaltrustregionprox_apply_modification(BundleDiagonalTrustRegionProx* self, const GroundsetModification* gsmdf) {
  return self->apply_modification(*gsmdf);
}

dll BundleProxObject* cb_bundlediagonaltrustregionprox_projected_clone(BundleDiagonalTrustRegionProx* self, const Indexmatrix* indices) {
  return self->projected_clone(*indices);
}

dll int cb_bundlediagonaltrustregionprox_supports_diagonal_bounds_scaling(const BundleDiagonalTrustRegionProx* self) {
  return self->supports_diagonal_bounds_scaling();
}

dll int cb_bundlediagonaltrustregionprox_diagonal_bounds_scaling_update(BundleDiagonalTrustRegionProx* self, const Matrix* D_update) {
  return self->diagonal_bounds_scaling_update(*D_update);
}

dll int cb_bundlediagonaltrustregionprox_supports_lowrank_variable_metric(const BundleDiagonalTrustRegionProx* self) {
  return self->supports_lowrank_variable_metric();
}

dll int cb_bundlediagonaltrustregionprox_supports_diagonal_variable_metric(const BundleDiagonalTrustRegionProx* self) {
  return self->supports_diagonal_variable_metric();
}

dll int cb_bundlediagonaltrustregionprox_apply_variable_metric(BundleDiagonalTrustRegionProx* self, VariableMetricModel* groundset, VariableMetricModel* model, const Matrix* aggr, Integer y_id, const Matrix* y, int descent_step, Real* current_weight, Real model_maxviol, const Indexmatrix* new_indices = 0) {
  return self->apply_variable_metric(groundset, model, *aggr, y_id, *y, (bool)descent_step, *current_weight, model_maxviol, new_indices);
}

dll int cb_bundlediagonaltrustregionprox_add_variable_metric(BundleDiagonalTrustRegionProx* self, Matrix* diagH, Matrix* vecH) {
  return self->add_variable_metric(*diagH, *vecH);
}

dll int cb_bundlediagonaltrustregionprox_push_aft(BundleDiagonalTrustRegionProx* self, const AffineFunctionTransformation* aft) {
  return self->push_aft(aft);
}

dll int cb_bundlediagonaltrustregionprox_pop_aft(BundleDiagonalTrustRegionProx* self) {
  return self->pop_aft();
}

dll int cb_bundlediagonaltrustregionprox_mfile_data(const BundleDiagonalTrustRegionProx* self) {
  return self->mfile_data(std::cout);
}

